docs/01-tooling/03-fast-and-isolated-JS-unit-tests.md

```diff
@@ -2,10 +2,159 @@
+`mw.*` functions or helpers, so we created a [`stubs.js`][stubs] where we
+created a few stub helpers for the tests.
+
+We also kept a couple of tests as integration tests in `tests/qunit`, but
+eventually we did some work to refactor the code and made unit tests for the new
+code, so we got rid of the integration tests in MediaWiki entirely.
+
+With this setup, tests run quite fast, and it is feasible (and we do) run them
+in watch mode while developing, giving you fast feedback and running your code
+on save, all from the terminal/editor.
+
+The environment doesn't have any global state, or implicit knowledge of
+MediaWiki, which forces us to write properly isolated tests that don't rely on
+implicit behavior or state.
+
+Finally, the move to a Node.js-based toolchain means we are easily able to
+leverage other great open source tools without much fuss, for example, for code
+coverage. We added another script -[`coverage`][coverage]- which just run the
+test command, but with the code coverage tool [`istanbul`][istanbul] first, and
+just like that we got back coverage reports for the frontend code.
+
+We recommend this approach for others wanting to improve how they test, and
+would be happy to help you figure out if this approach would work for you. For
+example, you can use this CLI runner, even if your JS sources just use globals
+instead of common.js or ES modules.
+
 ### Problems
+
+Overall, the move has gone great and we don't many issues to report.
+
+When migrating existing tests, it is sometimes a bit tricky to figure out how to
+move them to the isolated environment, since most of the [MediaWiki JS
+library][mw] is not available as an npm package, so in some occasions we had to
+restructure the code a bit to not implicitly assume as much of MediaWiki being
+available, and other times we had to set up some stubs for the tests to run
+well. This had the added benefit that the dependency on MediaWiki core libraries
+is explicit in the tests, so we should notice when adding new dependencies or
+changing them because of the failing tests, keeping the behavior and
+dependencies explicit.
+
+Another extra thing that we have been doing has been maintaining
+[`mw-node-qunit`][mw-node-qunit], which has taken a bit of additional time.
+Making sure our wrapper works well with qunitjs, and updating the dependency
+versions to not fall behind and leverage improvements on the libraries.
+
+We also expect some more work on the wrapper as [QUnit Plugins (or "Setup Step")
+in CLI \#1221][1221] gets worked on, or when [CLI: Implement --require option
+\#1271][1271] gets released, to simplify our implementation and be able to
+leverage the superior QUnit.js tap output and the rest of the CLI options like
+filters. The good thing is that the CLI wrapper should be useful for any
+Wikimedia projects as is.
+
+We will also be looking into moving the repository to the Wikimedia organization
+in github if other teams or projects adopt this testing strategy.
+
```

Comment:
It's a bit of downer closing with a problems section for something that worked out well. I'm also more interested in approaching shortcomings from the perspective of "how can we can do better" rather than "this was kind of broken." Consider adding a road ahead section here speculating on plans for the future (you talked about moving the rmw-node-qunit, for example), improvements, how we might apply lessons learned to MobileFrontend / MinervaNeue, things you ran out of time for in Popups that you want to do, etc. Maybe that's just changing this section's title and fleshing out some of the future plans and where you see us going or maybe that's adding a new section here at the end.

Reply:
Added the Conclusions section, I forgot!! I'll ping you when I commit to see what you think.